File: plugins/collector/event/apps/service/ntp.go
package service

import (
	"collector/cache/process"
	"collector/event/apps"
	"regexp"
	"strings"
)

type Ntp struct {
	version string
	reg     *regexp.Regexp
}

func (Ntp) Name() string { return "ntp" }

func (Ntp) Type() string { return "service" }

func (n Ntp) Version() string { return n.version }

func (Ntp) Match(p *process.Process) bool { return p.Name == "ntpd" }

func (n *Ntp) Run(p *process.Process) (m map[string]string, err error) {
	if n.reg == nil {
		n.reg = regexp.MustCompile(`ntpd (\d+\.)+\d+`)
	}
	result, err := apps.Execute(p, "--version")
	str := n.reg.FindString(result)
	if str == "" {
		err = apps.ErrVersionNotFound
		return
	}
	n.version = strings.TrimPrefix(str, "ntpd ")
	return
}

func init() { apps.Regist(&Ntp{}) }

File: plugins/collector/src/event/apps/service/ntp.rs
use crate::event::apps::{execute, find_version, AppProc, IApp};
use std::collections::HashMap;

pub struct Ntp {
    version: String,
}

impl Ntp {
    pub fn new() -> Self { Ntp { version: String::new() } }
}

impl IApp for Ntp {
    fn name(&self)     -> &'static str { "ntp" }
    fn app_type(&self) -> &'static str { "service" }
    fn version(&self)  -> &str         { &self.version }

    fn matches(&mut self, p: &AppProc) -> bool { p.name == "ntpd" }

    fn run(&mut self, p: &AppProc) -> anyhow::Result<HashMap<String, String>> {
        let out = execute(p, &["--version"])?;
        // "ntpd 4.2.8p15@1.3728-o …"
        self.version = find_version(&out);
        Ok(HashMap::new())
    }
}
